### ISSUE TITLE
`InvalidType` with Same-Named Parameters in Different Factories while using `readValue` for Deserialization

### ISSUE THREAD

**Issue by Vedsaga** [rrousselGit/freezed#1034]
**Describe the bug**
In Dart's freezed package, a type inference issue occurs when multiple factory constructors in a Freezed class have parameters with identical names and datatypes using a custom JSON deserialization method. The generated code mistakenly interprets these parameters as InvalidType. This problem is not present when parameter names differ across factories, despite having the same datatype and custom deserialization method.

**To Reproduce**

Copy Paste the below code in tow different file,

`metadata_model.dart`

```dart
// ignore_for_file: invalid_annotation_target
import 'package:freezed_annotation/freezed_annotation.dart';
import 'package:mytrailpals/common/common.dart';

part 'metadata_model.freezed.dart';
part 'metadata_model.g.dart';

@Freezed(fromJson: true, toJson: true)
class MetadataModel with _$MetadataModel {
  const MetadataModel._();

  const factory MetadataModel.view({
     required String id,
  }) = ViewMetadataModel;

  factory MetadataModel.fromJson(Map<String, dynamic> json) =>
      _$MetadataModelFromJson(json);

  @override
  Map<String, dynamic> toJson() {
    return switch (this) {
      ViewMetadataModel() =>
        _$$ViewMetadataModelImplToJson(this as _$ViewMetadataModelImpl),
      MetadataModel() => {},
    };
  }

  static ViewMetadataModel fromJsonReadValue(
    Map<dynamic, dynamic> json,
    String keyName,
  ) {
    if (json is! Map<String, dynamic>) {
      throw Exception('Json is not a Map<String, dynamic>');
    }
    return MetadataModel.fromJson(json) as ViewMetadataModel;
  }
}
```

```dart
// Demonstrating the issue with identical parameter names in two factories
// ignore_for_file: invalid_annotation_target
import 'package:freezed_annotation/freezed_annotation.dart';
import 'package:mytrailpals/common/common.dart';

part 'file_metadata_model.freezed.dart';
part 'file_metadata_model.g.dart';

@Freezed(fromJson: true, toJson: true)
class FileMetadataModel with _$FileMetadataModel {
  const FileMetadataModel._();

  const factory FileMetadataModel.view({
    @JsonKey(
      includeToJson: false,
      readValue: MetadataModel.fromJsonReadValue,
    )
    required ViewMetadataModel metaData,

  }) = ViewFileMetadataModel;

  const factory FileMetadataModel.createRemote({
    @JsonKey(
      includeToJson: false,
      readValue: MetadataModel.fromJsonReadValue,
    )
    required ViewMetadataModel metaData,
  }) = CreateRemoteFileMetadataModel;

  factory FileMetadataModel.fromJson(Map<String, dynamic> json) =>
      _$FileMetadataModelFromJson(json);

  @override
  Map<String, dynamic> toJson() {
    return switch (this) {
      ViewFileMetadataModel() =>
        _$$ViewFileMetadataModelImplToJson(this as _$ViewFileMetadataModelImpl),
      CreateRemoteFileMetadataModel() =>
        _$$CreateRemoteFileMetadataModelImplToJson(
          this as _$CreateRemoteFileMetadataModelImpl,
        ),
      FileMetadataModel() => {}
    };
  }

  static ViewFileMetadataModel fromJsonReadValue(
    Map<dynamic, dynamic> json,
    String keyName,
  ) {
    if (json is! Map<String, dynamic>) {
      throw Exception('Json is not a Map<String, dynamic>');
    }
    return FileMetadataModel.fromJson(json) as ViewFileMetadataModel;
  }
}
}
```

We would see that, in generated `.freezed.dart` file ` if we search we will find `InvalidType`.  

```dart
/// @nodoc
mixin _$FileMetadataModel {
  @JsonKey(includeToJson: false, readValue: MetadataModel.fromJsonReadValue)
  InvalidType get metaData => throw _privateConstructorUsedError;
...
}

/// @nodoc
abstract class $FileMetadataModelCopyWith<$Res> {
  factory $FileMetadataModelCopyWith(
          FileMetadataModel value, $Res Function(FileMetadataModel) then) =
      _$FileMetadataModelCopyWithImpl<$Res, FileMetadataModel>;
  @useResult
  $Res call(
      {@JsonKey(
          includeToJson: false, readValue: MetadataModel.fromJsonReadValue)
      InvalidType metaData});
}

/// @nodoc
class _$FileMetadataModelCopyWithImpl<$Res, $Val extends FileMetadataModel>
    implements $FileMetadataModelCopyWith<$Res> {
  _$FileMetadataModelCopyWithImpl(this._value, this._then);

  // ignore: unused_field
  final $Val _value;
  // ignore: unused_field
  final $Res Function($Val) _then;

  @pragma('vm:prefer-inline')
  @override
  $Res call({
    Object? metaData = freezed,
  }) {
    return _then(_value.copyWith(
      metaData: freezed == metaData
          ? _value.metaData
          : metaData // ignore: cast_nullable_to_non_nullable
              as InvalidType,
    ) as $Val);
  }
}
```

**Expected behavior**
The expected behavior is for the generated code to accurately infer the parameter types across different factory constructors without resulting in `InvalidType` errors, just similar to how it happens for in built data-types ( String, int, etc.).

**Comment by rrousselGit** [comment]
That's a known limitation, and it won't be fixed until we have metaprogramming.
TL;DR, don't use the generated classes in other Freezed classes

Closing as there are already issues about it
